### PR TITLE
Hooks

### PR DESCRIPTION
**Pull request by stephanebisson**
https://phabricator.wikimedia.org/T269934

Provide custom events hooks to allow 3rd party to build their own instrumentation when a popup is open and then the user navigates to read an article on wiki.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -5,19 +5,32 @@ import { createTouchPopup } from './touchPopup'
 import { renderPreview, renderLoading, renderError, renderDisambiguation, renderOffline } from './preview'
 import { getWikipediaAttrFromUrl, isTouch, getDir, isOnline } from './utils'
 
+const invokeCallback = ( events, name, ...params ) => {
```
Comment (hueitan) [inline src/index.js:8]:
```suggestion
const invokeCallback = ( events, name, params ) => {
```

this should be just `params` as it is already an array

File: demo/articles/english.html [hunk 2]
```diff
@@ -55,7 +55,11 @@
         <script src="../wikipedia-preview.development.js"></script>
         <script>
             wikipediaPreview.init({
-                lang: 'en'
+                lang: 'en',
+                events: {
+                    onShow: console.log,
+                    onReadOnWiki: console.log
```
Comment (hueitan) [inline demo/articles/english.html:61]: onWikiRead?

**Review by hueitan — APPROVED** [review]
Looks good now

File: src/index.js [hunk 3]
```diff
@@ -5,19 +5,32 @@ import { createTouchPopup } from './touchPopup'
 import { renderPreview, renderLoading, renderError, renderDisambiguation, renderOffline } from './preview'
 import { getWikipediaAttrFromUrl, isTouch, getDir, isOnline } from './utils'
 
+const invokeCallback = ( events, name, params ) => {
+	const callback = events && events[ name ]
+	if ( callback instanceof Function ) {
+		try {
+			callback.apply( null, params )
+		} catch ( e ) {
+			// we don't care about buggy 3rd party functions
```
Comment (medied) [inline src/index.js:14]: Would the errors be accessible to them? What if we at least `console.log` them in case they care?

File: src/index.js [hunk 4]
```diff
@@ -11,7 +11,7 @@ const invokeCallback = ( events, name, params ) => {
 		try {
 			callback.apply( null, params )
 		} catch ( e ) {
-			// we don't care about buggy 3rd party functions
+			console.log('Error invoking Wikipedia Preview custom callback' e )
```
Comment (jpita) [inline src/index.js:14]: the build is failing because of the `e`  here, should this be fixed?

**Comment by jpita** [comment]
issue found:
Steps
1-I move the mouse over a word 
2-move the mouse to the popup
3-move to the mouse to the word again

* Every time I repeat the steps 2 and 3 a new message appears. We should send only one message per "popup open", not when we over the mouse on the word

**Comment by hueitan** [comment]
> issue found:
> Steps
> 1-I move the mouse over a word
> 2-move the mouse to the popup
> 3-move to the mouse to the word again
> 
> * Every time I repeat the steps 2 and 3 a new message appears. We should send only one message per "popup open", not when we over the mouse on the word

so, this "bug" ticket exist here https://phabricator.wikimedia.org/T259490

**Comment by jpita** [comment]
> so, this "bug" ticket exist here https://phabricator.wikimedia.org/T259490
can be fixed here?

**Pull request merged**